rules-reviewed/camel3/camel2/xml-removed-components.windup.groovy

```diff
@@ -0,0 +1,82 @@
+package camel3.camel2
+
+import org.apache.commons.lang3.StringUtils
+import org.jboss.windup.config.GraphRewrite
+import org.jboss.windup.config.metadata.TechnologyReference
+import org.jboss.windup.config.operation.iteration.AbstractIterationOperation
+import org.jboss.windup.config.query.Query
+import org.jboss.windup.config.query.QueryPropertyComparisonType
+import org.jboss.windup.graph.model.resource.FileModel
+import org.jboss.windup.reporting.category.IssueCategory
+import org.jboss.windup.reporting.category.IssueCategoryRegistry
+import org.jboss.windup.reporting.config.Hint
+import org.jboss.windup.reporting.config.Link
+import org.jboss.windup.rules.apps.xml.condition.XmlFile
+import org.jboss.windup.rules.apps.xml.model.XmlFileModel
+import org.jboss.windup.rules.apps.xml.model.XmlTypeReferenceModel
+import org.jboss.windup.rules.files.condition.FileContent
+import org.ocpsoft.rewrite.config.Condition
+import org.ocpsoft.rewrite.context.EvaluationContext
+
+import java.util.function.BiFunction
+import java.util.function.Function
+
+final String FROM_XML_FILES_IN_PROJECT = "xmlFilesInProject"
+final String FROM_FILES_IN_PROJECT = "filesInProject"
+final Set<String> componentsAdded = ["aws-cw", "aws-ddb", "aws-ec2", "aws-iam", "aws-kinesis", "aws-kms", "aws-lambda", "aws-mq", "aws-s3", "aws-sdb", "aws-ses", "aws-sns", "aws-sqs", "aws-swf", "aws-xray"]
+
+final IssueCategory mandatoryIssueCategory = new IssueCategoryRegistry().getByID(IssueCategoryRegistry.MANDATORY)
+final Link awsLink = Link.to("Camel 3 - Migration Guide: AWS", "https://camel.apache.org/manual/latest/camel-3-migration-guide.html#_aws")
+
+final BiFunction<String, String, Boolean> isThereArtifactId = { String xmlDependenciesBlock, String component -> 
+    xmlDependenciesBlock.contains("<artifactId>camel-$component</artifactId>" as CharSequence) }
+
+final BiFunction<String, String, Condition> xmlCondition = { String component, String namespace -> 
+    XmlFile.from(FROM_XML_FILES_IN_PROJECT)
+            .matchesXpath("//*/c:route/*[starts-with(@uri, '$component:')]")
+            .namespace("c", namespace)
+            .as("$component-$namespace")
+}
+
+final Function<String, Condition> javaCondition = { String component ->
+    FileContent.from(FROM_FILES_IN_PROJECT)
+        .matches("(\"$component:")
+        .inFileNamed("{*}.java")
+        .as("$component-java")
+}
+
+ruleSet("xml-removed-components-groovy")
+    .addSourceTechnology(new TechnologyReference("camel", "[2,3)"))
+    .addTargetTechnology(new TechnologyReference("camel", "[3,)"))
+    .addRule()
+    .when(
+        XmlFile.matchesXpath("/m:project/m:dependencies[m:dependency/m:groupId/text() = 'org.apache.camel' and m:dependency/m:artifactId/text() = 'camel-aws']")
```

Comment:
If the application being migrated is using Spring Boot, the artifact used in the pom will be `camel-aws-starter` so it won't fit this requirement. Or it could include another dependency that transitively brings the dependency.
Instead of checking on the pom file , we could check that an artifact is present on the classpath `camel-aws`.
wdyt ?

Reply:
If i don't misunderstand you, i think your suggestion won't work.
`making the artifact query ["isThereArtifactId" my understanding] the main and not search on the pom` won't work because `isThereArtifactId` filter is a "negative" condition that searches for something missing: whatever application with nothing about Camel AWS would trigger such a condition.
So, if used as the main condition, this would have a negative impact on performances compared to have the current condition that restricts the applications analyzed to only the one related to Camel AWS.

On the first part of your comment, i think the original XML rule was based on https://camel.apache.org/manual/latest/camel-3-migration-guide.html#_aws 
@johnpoth and @mmelko i think @jonathanvila raised a good point about Spring Boot use case for Camel. The Camel migration guide doesn't mention it explicitly: is it worth adding it to the guide? If yes, then we could raise a Jira to improve also this rule.

Reply:
My suggestion would be to transform the POM dependency check to a artifact check.  
In this case it is checking that the POM is including a dependency org.apache.camel:camel-aws.  
So instead of that it could be checking that the artifact org.apache.camel:camel-aws exists in the fat jar.

Reply:
@mrizzi @jonathanvila  good catch! Yes, the same needs to be done for camel-aws-starter artifact that has been split into multiple artifacts. I'll update the website and create a separate JIRA in WINDUP. This is also relevant to the other Camel components that were removed/renamed/split (cc @mmelko ).

I think what @jonathanvila is suggesting is for windup to search the project's dependency tree and not just parse the raw pom.xml

Reply:
I think removing the current POM dependency check is not suitable because it would prevent the rule to work with source code projects (where you have the `pom.xml` file and not a compiled fat-jar to analyze).

The solution I can propose is adding the suggested check for the artifact in the fat-jar in an `or` condition with the current condition (to preserve the source mode use case) and then changing  `XmlTypeReferenceModel` payload to become "just" a `FileLocationModel` payload that would preserve the hint in the specific location within the file identified in the `when` condition.

Reply:
@johnpoth no need for you to create the issue in windup's Jira, i already did it [WINDUPRULE-592](https://issues.redhat.com/browse/WINDUPRULE-592)